---
title: "Weighted Sliced Inverse Regression (WSIR): supervised dimension reduction for spatial transcriptomics data"
author:
  - Max Woollard
  - Linh Nghiem
  - Shila Ghazanfar
  - Pratibha Panwar
output:
  BiocStyle::html_document:
    toc_float: true
  BiocStyle::pdf_document: default
package: wSIR
vignette: |
  %\VignetteIndexEntry{WSIR supervised dimension reduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = FALSE
)
```

```{r, message=FALSE, warning=FALSE}
library(wSIR) # package itself
library(magrittr) # for %>% 
library(ggplot2) # for ggplot
library(Rfast) # for distance correlation
library(doBy) # for which.maxn
library(vctrs) # for vec_rep_each
library(umap) # for umap
library(BiocParallel) # for parallel computing in exploreWSIRParams with BiocParallel

# Packages needed to download data
#library(scran) # for logNormCounts
#library(MouseGastrulationData) # to download the data for this vignette
```

# Introduction

Weighted Sliced Inverse Regression (wSIR) is a supervised dimension reduction technique for spatial transcriptomics data. This method uses each cell's gene expression values as covariates and spatial position as the response. This allows us to create a low-dimensional representation of the gene expression data that retains the information about spatial patterns that is present in the gene expression data. The resulting mapping from gene expression data to a spatially aware low-dimensional embedding can be used to project new single-cell gene expression data into a low-dimensional space which preserves the ability to predict each cell's spatial location from its low-dimensional embedding. 

## Method Overview

wSIR is an extension of the supervised dimension reduction technique of Sliced Inverse Regression (SIR). 

SIR is an existing supervised dimension reduction method which works by grouping together the observations with similar values for the response. For spatial transcriptomics data, this means grouping all the cells into a certain number of tiles based on their spatial position. For example, if we use 4 tiles, then the cells in the top right quadrant of the image/tissue go to one group, those in the top left to another, and so on. Each of those groups is then summarised by averaging the expression level of all cells in each group for each of the genes. From there, eigendecomposition is performed on the resulting matrix of tile-gene means, then returning the SIR directions and SIR scores.

The motivation behind wSIR is that SIR only uses each cell's spatial position when we are separating the cells into the given number of groups/tiles. Once those groups are created, we lose the fact that some groups may be more spatially related (if they come from adjacent tiles) than other groups (if they come from opposite sides of the tissue). wSIR uses a weight matrix to incorporate the spatial correlation between all pairs of cells in the SIR algorithm. This matrix has dimension H*H, where H is the number of tiles, and the (i,j)th entry represents the distance between tiles i and j. This metrix is incorporated into the eigendeomposition step. The wSIR output has the same structure as the SIR output. 

## Vignette Goals

In this vignette, we will demonstrate how to use WSIR to obtain a low-dimensional embedding of gene expression data. We will then explore this embedding using the package's built-in functions. However, this low-dimensional matrix is more importantly used for your own downstream tasks that would benefit from a lower-dimensional representation of gene expression data that preserves information about each cell's spatial location. We perform some basic downstream analysis to demonstrate the practicality of wSIR.

# Data

We use data from https://www.nature.com/articles/s41587-021-01006-2 . I have commented out the code to download it yourself (which requires packages MouseGastrulationData and scran). For this vignette and the examples in the .R function files, we simply load this data that has already been saved at data/MouseData.Rdata.

Note: there is a BioConductor limit on the size of data that can be used for examples and vignettes. To stay under that, we randomly sample 20% of the cells from each of the three samples. 

```{r}
#seqfish_data_sample1 <- LohoffSeqFISHData(samples = 1)
#seqfish_data_sample1 = logNormCounts(seqfish_data_sample1) # log transform variance stabilising 
#rownames(seqfish_data_sample1) <- rowData(seqfish_data_sample1)[,"SYMBOL"] # change rownames to gene symbols that are consistent across Google 
#sample1_exprs = t(assay(seqfish_data_sample1, "logcounts")) # extract matrix of gene expressions 
#sample1_coords = spatialCoords(seqfish_data_sample1)[,1:2] %>% as.data.frame()
#colnames(sample1_coords) = c("x", "y")

#seqfish_data_sample2 <- LohoffSeqFISHData(samples = 2)
#seqfish_data_sample2 = logNormCounts(seqfish_data_sample2) # log transform variance stabilising 
#rownames(seqfish_data_sample2) <- rowData(seqfish_data_sample2)[,"SYMBOL"] # change rownames to gene symbols that are consistent across Google 
#sample2_exprs = t(assay(seqfish_data_sample2, "logcounts")) # extract matrix of gene expressions 
#sample2_coords = spatialCoords(seqfish_data_sample2)[,1:2] %>% as.data.frame()
#colnames(sample2_coords) = c("x", "y")

#seqfish_data_sample3 <- LohoffSeqFISHData(samples = 3)
#seqfish_data_sample3 = logNormCounts(seqfish_data_sample3) # log transform variance stabilising 
#rownames(seqfish_data_sample3) <- rowData(seqfish_data_sample3)[,"SYMBOL"] # change rownames to gene symbols that are consistent across Google 
#sample3_exprs = t(assay(seqfish_data_sample3, "logcounts")) # extract matrix of gene expressions 
#sample3_coords = spatialCoords(seqfish_data_sample3)[,1:2] %>% as.data.frame()
#colnames(sample3_coords) = c("x", "y")

#set.seed(2024)
#keep1 = sample(c(TRUE, FALSE), nrow(sample1_exprs), replace = TRUE, prob = c(0.2, 0.8))
#keep2 = sample(c(TRUE, FALSE), nrow(sample2_exprs), replace = TRUE, prob = c(0.2, 0.8))
#keep3 = sample(c(TRUE, FALSE), nrow(sample3_exprs), replace = TRUE, prob = c(0.2, 0.8))

#sample1_exprs = sample1_exprs[keep1,]
#sample1_coords = sample1_coords[keep1,]
#sample2_exprs = sample2_exprs[keep2,]
#sample2_coords = sample2_coords[keep2,]
#sample3_exprs = sample3_exprs[keep3,]
#sample3_coords = sample3_coords[keep3,]

#save(sample1_exprs, sample1_coords, sample2_exprs, sample2_coords, sample3_exprs, sample3_coords, file = "MouseData.RData", compress = "xz")

data(MouseData)
```

# Supervised dimension reduction with WSIR

Tune parameters to find optimal values for parameters slices, alpha

```{r}
optim_obj = exploreWSIRParams(exprs = sample1_exprs, 
                           coords = sample1_coords, 
                           alpha_vals = c(0,2,4,8), 
                           slice_vals = c(3,6,10))
optim_obj$plot
```

Perform WSIR

```{r}
wsir_obj = wSIR(X = sample1_exprs, 
                coords = sample1_coords, 
                slices = optim_obj$best_slices, 
                alpha = optim_obj$best_alpha,
                optim_params = FALSE) # create wsir object using optimal values for slices, alpha
```

Observe the genes that contribute the most to WSIR1 (first direction of WSIR) using top_genes function

```{r}
top_genes_obj = findTopGenes(WSIR = wsir_obj, highest = 8) # create top genes object
top_genes_plot = top_genes_obj$plot # select plot
top_genes_plot # print plot
```

Visualise the WSIR directions using visualise_wsir function

```{r}
vis_obj = visualiseWSIRDirections(coords = sample1_coords, WSIR = wsir_obj, dirs = 8) # create visualisations
vis_obj
```

Visualise UMAP plots of low-dimensional representation of sample1_exprs after WSIR, colouring points by the 4 genes that contribute the most to WSIR1 using vis_umap function

```{r}
umap_coords = generateUmapFromWSIR(WSIR = wsir_obj)
umap_plots = plotUmapFromWSIR(exprs = sample1_exprs,
                              umap_coords = umap_coords,
                              highest_genes = top_genes_obj,
                              n_genes = 6)
umap_plots
```

Projecting sample2_exprs into low-dimensional space using project_wSIR function. This low-dimensional embedding retains the ability to predict a cell's spatial location that was already present in the gene expression data.

```{r}
sample2_low_dim_exprs = projectWSIR(wsir = wsir_obj, newdata = sample2_exprs)
dim(sample2_low_dim_exprs)
```

<details>
  <summary>**Session Info**</summary>
  
```{r}
sessionInfo()
```

</details>
