<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Weighted Sliced Inverse Regression (WSIR) • wSIR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Weighted Sliced Inverse Regression (WSIR)">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">wSIR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/wSIR_vignette.html">Weighted Sliced Inverse Regression (WSIR)</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sydneybiox/wSIR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Weighted Sliced Inverse Regression (WSIR)</h1>
                        <h4 data-toc-skip class="author">Max
Woollard</h4>
                        <h4 data-toc-skip class="author">Pratibha
Panwar</h4>
                        <h4 data-toc-skip class="author">Linh
Nghiem</h4>
                        <h4 data-toc-skip class="author">Shila
Ghazanfar</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sydneybiox/wSIR/blob/main/vignettes/wSIR_vignette.Rmd" class="external-link"><code>vignettes/wSIR_vignette.Rmd</code></a></small>
      <div class="d-none name"><code>wSIR_vignette.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/wSIR">wSIR</a></span><span class="op">)</span> <span class="co"># package itself</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span> <span class="co"># for %&gt;% </span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span> <span class="co"># for ggplot</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RfastOfficial/Rfast" class="external-link">Rfast</a></span><span class="op">)</span> <span class="co"># for distance correlation</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">doBy</span><span class="op">)</span> <span class="co"># for which.maxn</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://vctrs.r-lib.org/" class="external-link">vctrs</a></span><span class="op">)</span> <span class="co"># for vec_rep_each</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tkonopka/umap" class="external-link">umap</a></span><span class="op">)</span> <span class="co"># for umap</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.stats.ox.ac.uk/pub/MASS4/" class="external-link">class</a></span><span class="op">)</span> <span class="co"># for example wSIR application</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Weighted Sliced Inverse Regression (wSIR) is a supervised dimension
reduction technique for spatial transcriptomics data. This method uses
each cell’s gene expression values as covariates and spatial position as
the response. This allows us to create a low-dimensional representation
of the gene expression data that retains the information about spatial
patterns that is present in the gene expression data. The resulting
mapping from gene expression data to a spatially aware low-dimensional
embedding can be used to project new single-cell gene expression data
into a low-dimensional space which preserves the ability to predict each
cell’s spatial location from its low-dimensional embedding.</p>
<div class="section level3">
<h3 id="method-overview">Method Overview<a class="anchor" aria-label="anchor" href="#method-overview"></a>
</h3>
<p>wSIR is an extension of the supervised dimension reduction technique
of Sliced Inverse Regression (SIR).</p>
<p>SIR is an existing supervised dimension reduction method which works
by grouping together the observations with similar values for the
response. For spatial transcriptomics data, this means grouping all the
cells into a certain number of tiles based on their spatial position.
For example, if we use 4 tiles, then the cells in the top right quadrant
of the image/tissue go to one group, those in the top left to another,
and so on. Each of those groups is then summarised by averaging the
expression level of all cells in each group for each of the genes. From
there, eigendecomposition is performed on the resulting matrix of
tile-gene means, then returning the SIR directions and SIR scores.</p>
<p>The motivation behind wSIR is that SIR only uses each cell’s spatial
position when we are separating the cells into the given number of
groups/tiles. Once those groups are created, we lose the fact that some
groups may be more spatially related (if they come from adjacent tiles)
than other groups (if they come from opposite sides of the tissue). wSIR
uses a weight matrix to incorporate the spatial correlation between all
pairs of cells in the SIR algorithm. This matrix has dimension H*H,
where H is the number of tiles, and the (i,j)th entry represents the
distance between tiles i and j. This matrix is incorporated into the
eigendeomposition step. The wSIR output has the same structure as the
SIR output.</p>
</div>
<div class="section level3">
<h3 id="vignette-overview">Vignette Overview<a class="anchor" aria-label="anchor" href="#vignette-overview"></a>
</h3>
<p>In this vignette, we will demonstrate how to use WSIR to obtain a
low-dimensional embedding of gene expression data. We will then explore
this embedding using the package’s built-in functions. However, this
low-dimensional matrix is more importantly used for your own downstream
tasks that would benefit from a lower-dimensional representation of gene
expression data that preserves information about each cell’s spatial
location. We perform some basic downstream analysis to demonstrate the
practicality of wSIR.</p>
</div>
</div>
<div class="section level2">
<h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h2>
<p>We use data from <a href="https://www.nature.com/articles/s41587-021-01006-2" class="external-link">Lohoff et al,
2021</a>. The code below would allow you to download the mouse data
yourself using the MouseGastrulationData and scran R packages. We
randomly sample 20% of the cells from each of the three biological
replicate samples.</p>
<pre><code><span><span class="co"># Packages needed to download data</span></span>
<span><span class="co">#library(scran) # for logNormCounts</span></span>
<span><span class="co">#library(MouseGastrulationData) # to download the data for this vignette</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seqfish_data_sample1</span> <span class="op">&lt;-</span> <span class="fu">LohoffSeqFISHData</span><span class="op">(</span>samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">seqfish_data_sample1</span> <span class="op">=</span> <span class="fu">logNormCounts</span><span class="op">(</span><span class="va">seqfish_data_sample1</span><span class="op">)</span> <span class="co"># log transform variance stabilising</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">seqfish_data_sample1</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">seqfish_data_sample1</span><span class="op">)</span><span class="op">[</span>,<span class="st">"SYMBOL"</span><span class="op">]</span> <span class="co"># change rownames to gene symbols that are consistent</span></span>
<span><span class="va">sample1_exprs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">seqfish_data_sample1</span>, <span class="st">"logcounts"</span><span class="op">)</span><span class="op">)</span> <span class="co"># extract matrix of gene expressions</span></span>
<span><span class="va">sample1_coords</span> <span class="op">=</span> <span class="fu">spatialCoords</span><span class="op">(</span><span class="va">seqfish_data_sample1</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sample1_coords</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seqfish_data_sample2</span> <span class="op">&lt;-</span> <span class="fu">LohoffSeqFISHData</span><span class="op">(</span>samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">seqfish_data_sample2</span> <span class="op">=</span> <span class="fu">logNormCounts</span><span class="op">(</span><span class="va">seqfish_data_sample2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">seqfish_data_sample2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">seqfish_data_sample2</span><span class="op">)</span><span class="op">[</span>,<span class="st">"SYMBOL"</span><span class="op">]</span></span>
<span><span class="va">sample2_exprs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">seqfish_data_sample2</span>, <span class="st">"logcounts"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sample2_coords</span> <span class="op">=</span> <span class="fu">spatialCoords</span><span class="op">(</span><span class="va">seqfish_data_sample2</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sample2_coords</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seqfish_data_sample3</span> <span class="op">&lt;-</span> <span class="fu">LohoffSeqFISHData</span><span class="op">(</span>samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">seqfish_data_sample3</span> <span class="op">=</span> <span class="fu">logNormCounts</span><span class="op">(</span><span class="va">seqfish_data_sample3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">seqfish_data_sample3</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">seqfish_data_sample3</span><span class="op">)</span><span class="op">[</span>,<span class="st">"SYMBOL"</span><span class="op">]</span></span>
<span><span class="va">sample3_exprs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">seqfish_data_sample3</span>, <span class="st">"logcounts"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sample3_coords</span> <span class="op">=</span> <span class="fu">spatialCoords</span><span class="op">(</span><span class="va">seqfish_data_sample3</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sample3_coords</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">keep1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sample1_exprs</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span>, prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">0.8</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">keep2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sample2_exprs</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span>, prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">0.8</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">keep3</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sample3_exprs</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span>, prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">0.8</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sample1_exprs</span> <span class="op">=</span> <span class="va">sample1_exprs</span><span class="op">[</span><span class="va">keep1</span>,<span class="op">]</span></span>
<span><span class="va">sample1_coords</span> <span class="op">=</span> <span class="va">sample1_coords</span><span class="op">[</span><span class="va">keep1</span>,<span class="op">]</span></span>
<span><span class="va">sample2_exprs</span> <span class="op">=</span> <span class="va">sample2_exprs</span><span class="op">[</span><span class="va">keep2</span>,<span class="op">]</span></span>
<span><span class="va">sample2_coords</span> <span class="op">=</span> <span class="va">sample2_coords</span><span class="op">[</span><span class="va">keep2</span>,<span class="op">]</span></span>
<span><span class="va">sample3_exprs</span> <span class="op">=</span> <span class="va">sample3_exprs</span><span class="op">[</span><span class="va">keep3</span>,<span class="op">]</span></span>
<span><span class="va">sample3_coords</span> <span class="op">=</span> <span class="va">sample3_coords</span><span class="op">[</span><span class="va">keep3</span>,<span class="op">]</span></span>
<span></span>
<span><span class="va">sample1_cell_types</span> <span class="op">=</span> <span class="va">seqfish_data_sample1</span><span class="op">$</span><span class="va">celltype</span><span class="op">[</span><span class="va">keep1</span><span class="op">]</span></span>
<span><span class="va">sample2_cell_types</span> <span class="op">=</span> <span class="va">seqfish_data_sample2</span><span class="op">$</span><span class="va">celltype</span><span class="op">[</span><span class="va">keep2</span><span class="op">]</span></span>
<span><span class="va">sample3_cell_types</span> <span class="op">=</span> <span class="va">seqfish_data_sample3</span><span class="op">$</span><span class="va">celltype</span><span class="op">[</span><span class="va">keep3</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">sample1_exprs</span>, <span class="va">sample1_coords</span>, <span class="va">sample1_cell_types</span>, </span>
<span>     <span class="va">sample2_exprs</span>, <span class="va">sample2_coords</span>, <span class="va">sample2_cell_types</span>, </span>
<span>     <span class="va">sample3_exprs</span>, <span class="va">sample3_coords</span>, <span class="va">sample3_cell_types</span>, </span>
<span>     file <span class="op">=</span> <span class="st">"../data/MouseData.rda"</span>, compress <span class="op">=</span> <span class="st">"xz"</span><span class="op">)</span></span></code></pre>
<p>For this vignette and the examples for each function in wSIR, we
simply load this data that has already been saved at
<code>data/MouseData.rda</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">MouseData</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="supervised-dimension-reduction-with-wsir">Supervised dimension reduction with wSIR<a class="anchor" aria-label="anchor" href="#supervised-dimension-reduction-with-wsir"></a>
</h2>
<div class="section level3">
<h3 id="parameter-study">Parameter Study<a class="anchor" aria-label="anchor" href="#parameter-study"></a>
</h3>
<p>wSIR contains two parameters, <code>alpha</code> and
<code>slices</code>. Parameter <code>slices</code> is the number of
groups along each spatial axis into which we split the observations in
the wSIR algorithm. More slices means we could pick up more spatial
information in the gene expression data, but we risk overfitting on the
training set. Parameter <code>alpha</code> modifies the strength of the
weight matrix. <code>alpha = 0</code> is equivalent to no spatial
weighting, meaning the weight matrix becomes the identity matrix. This
is equivalent to SIR. Larger <code>alpha</code> values means there is
more weight given to spatial correlation rather than gene expression
differences alone in the computation of the wSIR directions.</p>
<p>The function <code>exploreWSIRParams</code> performs wSIR over many
choices of <code>alpha</code> and <code>slices</code> to identify the
most appropriate parameters moving forward. Here, we will use sample 1
only.</p>
<p>These parameters should both be tuned over some reasonable values.
The function <code>exploreWSIRParams</code> computes and visualises the
performance of wSIR for all given combinations of <code>slices</code>
and <code>alpha</code>. The performance is computed from the following
procedure: 1) For each combination of <code>slices</code> and
<code>alpha</code>, split the data into 50% train and 50% test (where
train and test halves both include gene expression data and cell
coordinates). 2) Perform wSIR on the training set using the current
combination of slices and alpha. 3) Project the gene expression data of
the testing set into low-dimensional space using the wSIR results from
step 2. 4) Evaluate wSIR’s performance by computing either the distance
correlation (“DC”, default), or the correlation of distances (“CD”)
between the projected gene expression data of the test set and the test
coordinates, according to parameter <code>metric</code>. 5) Repeat steps
2-4 until it has been done <code>nrep</code> times (<code>nrep</code> is
a parameter whose default is 5). Calculate the average metric value over
each of the <code>nrep</code> iterations for this combination of slices
and alpha. 6) Repeat steps 1-5 with all other combinations of
<code>slices</code> and <code>alpha</code> to obtain an average metric
value for each combination. 7) Return the combination of
<code>slices</code> and <code>alpha</code> with the highest average
metric value and display a plot showing the performance for every
combination of parameters.</p>
<p>Note: a key advantage of wSIR over SIR is parameter robustness. Here,
we can see that SIR’sperformance deteriorates as you use larger values
for <code>slices</code>. However, wSIR’s performance is relatively
stable as you vary both <code>slices</code> and <code>alpha</code>
across reasonable values (e.g. among the default values we optimise
over). In the following plot, the metric value becomes smaller as we
increase the number of slices for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi><mo>=</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\alpha = 0</annotation></semantics></math>
(which corresponds to SIR). Performance is stable for all non-zero alpha
and slice combinations (which correspond to wSIR).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">optim_obj</span> <span class="op">=</span> <span class="fu"><a href="../reference/exploreWSIRParams.html">exploreWSIRParams</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">sample1_exprs</span>, </span>
<span>                              coords <span class="op">=</span> <span class="va">sample1_coords</span>, </span>
<span>                              alpha_vals <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">.5</span>, <span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">8</span><span class="op">)</span>, </span>
<span>                              slice_vals <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fl">10</span>,<span class="fl">15</span><span class="op">)</span>,</span>
<span>                              nrep <span class="op">=</span> <span class="fl">20</span>, metric <span class="op">=</span> <span class="st">"CD"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span><span class="op">-</span><span class="va">a</span></span></code></pre></div>
<pre><code><span><span class="co">## Time difference of 3.80371 mins</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">optim_obj</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<p><img src="wSIR_vignette_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="wsir-computation">wSIR Computation<a class="anchor" aria-label="anchor" href="#wsir-computation"></a>
</h3>
<p>We next perform wSIR using the optimal parameter combination that we
found in the previous section. We use the gene expression matrix and
spatial coordinates from sample 1 here. This returns a list of results
with 5 (named) slots, whose details can be found at
<code><a href="../reference/wSIR.html">?wSIR::wSIR</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wsir_obj</span> <span class="op">=</span> <span class="fu"><a href="../reference/wSIR.html">wSIR</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">sample1_exprs</span>, </span>
<span>                coords <span class="op">=</span> <span class="va">sample1_coords</span>, </span>
<span>                slices <span class="op">=</span> <span class="va">optim_obj</span><span class="op">$</span><span class="va">best_slices</span>, </span>
<span>                alpha <span class="op">=</span> <span class="va">optim_obj</span><span class="op">$</span><span class="va">best_alpha</span>,</span>
<span>                optim_params <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">wsir_obj</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "scores"     "directions" "estd"       "W"          "evalues"</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="wsir-results-analysis">wSIR Results Analysis<a class="anchor" aria-label="anchor" href="#wsir-results-analysis"></a>
</h2>
<p>In this section, we will demonstrate how to use the built-in analysis
functions to better understand how wSIR creates a spatially-informed
low-dimensional embedding. These functions all use a wSIR result as an
input. Here, we use the output from the previous section, meaning we are
studying the result of performing wSIR on sample 1 only.</p>
<div class="section level3">
<h3 id="wsir-top-genes">wSIR Top Genes<a class="anchor" aria-label="anchor" href="#wsir-top-genes"></a>
</h3>
<p>The <code>findTopGenes</code> function finds and plots the genes with
highest loading in the specified wSIR directions (default is direction
1). If a gene has high loading (in terms of magnitude), it is more
important to the wSIR direction. Since the wSIR directions are designed
to retain information about each cell’s spatial position, the genes with
high loading should be spatially-related genes.</p>
<p>In the plot below, we can see which genes have the highest loading in
wSIR direction 1. This is useful as it gives us an intuition about how
wSIR creates the low-dimensional embedding. We can see that some of the
genes are known spatial genes (e.g. Cdx2, Hox-), which is what we would
expect to see.</p>
<p>We can simultaneously plot top genes for multiple directions, and
utilise the</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_genes_obj</span> <span class="op">=</span> <span class="fu"><a href="../reference/findTopGenes.html">findTopGenes</a></span><span class="op">(</span>WSIR <span class="op">=</span> <span class="va">wsir_obj</span>, highest <span class="op">=</span> <span class="fl">8</span><span class="op">)</span> <span class="co"># create top genes object</span></span>
<span><span class="va">top_genes_plot</span> <span class="op">=</span> <span class="va">top_genes_obj</span><span class="op">$</span><span class="va">plot</span> <span class="co"># select plot</span></span>
<span><span class="va">top_genes_plot</span> <span class="co"># print plot</span></span></code></pre></div>
<p><img src="wSIR_vignette_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_genes_obj</span> <span class="op">=</span> <span class="fu"><a href="../reference/findTopGenes.html">findTopGenes</a></span><span class="op">(</span>WSIR <span class="op">=</span> <span class="va">wsir_obj</span>, highest <span class="op">=</span> <span class="fl">8</span>, dirs <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">top_genes_plot</span> <span class="op">=</span> <span class="va">top_genes_obj</span><span class="op">$</span><span class="va">plot</span></span>
<span><span class="va">top_genes_plot</span></span></code></pre></div>
<p><img src="wSIR_vignette_files/figure-html/unnamed-chunk-6-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualising-wsir-scores">Visualising wSIR Scores<a class="anchor" aria-label="anchor" href="#visualising-wsir-scores"></a>
</h3>
<p>The <code>visualiseWSIRDirections</code> function plots each cell at
its spatial position, coloured by its value for each of the specified
wSIR columns. This gives us an understanding of what each column of the
low-dimensional embedding represents.</p>
<p>Below, we visualise the cells at their spatial positions, coloured by
each of the 5 wSIR directions The top left plot illustrates how, for
this example, wSIR direction 1 captures information about the “y”
spatial axis, since cells with higher “y” coordinate have low wSIR1
value, while cells with lower “y” coordinate have higher wSIR1 value.
wSIR2 is shown in the next plot over (the one titled “2”), and we can
see that wSIR column two appears to capture information about the “x”
spatial coordinate. The remaining three wSIR columns all contain
information about cell types, which we can tell by the regions of high
and low wSIR column values spread across the tissue.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vis_obj</span> <span class="op">=</span> <span class="fu"><a href="../reference/visualiseWSIRDirections.html">visualiseWSIRDirections</a></span><span class="op">(</span>coords <span class="op">=</span> <span class="va">sample1_coords</span>, WSIR <span class="op">=</span> <span class="va">wsir_obj</span>, dirs <span class="op">=</span> <span class="fl">8</span><span class="op">)</span> <span class="co"># create visualisations</span></span>
<span><span class="va">vis_obj</span></span></code></pre></div>
<p><img src="wSIR_vignette_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="umap-on-low-dimensional-embedding">UMAP on low-dimensional embedding<a class="anchor" aria-label="anchor" href="#umap-on-low-dimensional-embedding"></a>
</h3>
<p>The two functions <code>generateUmapFromWSIR</code> and
<code>plotUmapFromWSIR</code> create and display UMAP dimension
reduction calculated on the wSIR low-dimensional embedding. We can
colour the UMAP plot (where each point represents a cell) by its value
for various genes of interest. This visualises the structure of the wSIR
dimension reduction space, which is useful to gain more intuition about
what the space represents. Specifically, we can see if the wSIR space
contains neighbourhoods of high expression for specific genes, thus
better understanding how this space is made.</p>
<p>To specify which genes we would like to include, we can use the
output from the <code>findTopGenes</code> function from above, which
finds spatially-related genes by ranking those with the highest loading
in relevant wSIR directions. This output is then the value for the
<code>highest_genes</code> parameter. Otherwise, we could also specify
our own genes of interest if there are some specific genes we would like
to visualise. For example, if we wanted to visualise the expression
distribution for Cdx2 and Hoxb4, we could use
<code>genes = c("Cdx2", "Hoxb4")</code> as an argument in
<code>plotUmapFromWSIR</code> (and leave <code>highest_genes</code>
blank).</p>
<p>Below, we use the UMAP function to visualise the wSIR space computed
on the gene expression data from sample 1. We colour each cell by their
values for the 6 genes with highest value in wSIR direction 1 (as found
by the <code>findTopGenes</code> function previously). We can see that
for some of these genes, there are specific regions of high expression
in the UMAP plots, suggesting that the wSIR space separates cells based
on their expression for those genes.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">umap_coords</span> <span class="op">=</span> <span class="fu"><a href="../reference/generateUmapFromWSIR.html">generateUmapFromWSIR</a></span><span class="op">(</span>WSIR <span class="op">=</span> <span class="va">wsir_obj</span><span class="op">)</span></span>
<span><span class="va">umap_plots</span> <span class="op">=</span> <span class="fu"><a href="../reference/plotUmapFromWSIR.html">plotUmapFromWSIR</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">sample1_exprs</span>,</span>
<span>                              umap_coords <span class="op">=</span> <span class="va">umap_coords</span>,</span>
<span>                              highest_genes <span class="op">=</span> <span class="va">top_genes_obj</span>,</span>
<span>                              n_genes <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span><span class="va">umap_plots</span></span></code></pre></div>
<p><img src="wSIR_vignette_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="projection-of-new-data-with-wsir">Projection of new data with wSIR<a class="anchor" aria-label="anchor" href="#projection-of-new-data-with-wsir"></a>
</h2>
<p>A key functionality of the wSIR package is the ability to project new
single-cell data into the wSIR low-dimensional space. This will allow
for a low-dimensional representation of gene expression data that
contains information about each cell’s spatial position even though we
do not have access to the spatial coordinates for this new data. This
low-dimensional wSIR embedding would be especially useful for downstream
applications, like spatial alignment or spatial clustering (where we
don’t have spatial coordinates).</p>
<p>Here, we will demonstrate the steps for that, as well as a specific
application.</p>
<p>For each projection example, we will perform wSIR on a spatial
transcriptomics dataset which includes gene expression data and spatial
coordinates. We will then project a “single-cell” dataset, which only
contains the gene expression matrix, into wSIR low-dimensional
space.</p>
<div class="section level3">
<h3 id="single-sample-spatial-dataset-single-sample-single-cell-dataset">Single-sample spatial dataset, single-sample single-cell
dataset<a class="anchor" aria-label="anchor" href="#single-sample-spatial-dataset-single-sample-single-cell-dataset"></a>
</h3>
<p>Here, we demonstrate the steps to project a new sample single-cell
dataset into wSIR low-dimensional space having already performed wSIR on
the spatial transcriptomics dataset from the first sample.</p>
<p>We have already performed wSIR on sample 1, so here we project sample
2’s gene expression matrix into the wSIR low-dimensional space, which
will therefore have an ability to predict sample 2’s (unknown at this
stage) spatial locations.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample2_low_dim_exprs</span> <span class="op">=</span> <span class="fu"><a href="../reference/projectWSIR.html">projectWSIR</a></span><span class="op">(</span>wsir <span class="op">=</span> <span class="va">wsir_obj</span>, newdata <span class="op">=</span> <span class="va">sample2_exprs</span><span class="op">)</span></span></code></pre></div>
<p>Check the dimension of sample 2’s low-dimensional gene expression
data:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">sample2_low_dim_exprs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 2986   20</span></span></code></pre>
<p>Observe some of sample 2’s low-dimensional gene expression data:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sample2_low_dim_exprs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           [,1]       [,2]       [,3]       [,4]       [,5]       [,6]</span></span>
<span><span class="co">## [1,] 1.0930341 -0.1443871 -1.5090562  1.5873578 -1.3890866  0.3799610</span></span>
<span><span class="co">## [2,] 0.4680595  0.5327497 -0.3728135  1.0015734  0.8920196  0.1513758</span></span>
<span><span class="co">## [3,] 1.2711455 -0.7397591 -1.4112419 -0.8839003  0.2468790  0.3832880</span></span>
<span><span class="co">## [4,] 1.7531946 -0.1143880 -0.7064318 -0.3602051  0.4724328 -0.2961441</span></span>
<span><span class="co">## [5,] 2.2381875 -0.5924558 -0.3165188  0.2141844  0.7782123 -0.2621179</span></span>
<span><span class="co">## [6,] 1.3299499 -0.8290736  0.6474762  0.6397734  0.5299646 -0.9353155</span></span>
<span><span class="co">##            [,7]       [,8]        [,9]      [,10]       [,11]      [,12]</span></span>
<span><span class="co">## [1,] -3.4678998 0.56681244 -0.81551059  1.1852559 -0.73526941 -1.0785081</span></span>
<span><span class="co">## [2,] -1.0017729 1.95474647 -0.43692515 -0.1603876 -0.15958819 -0.4427703</span></span>
<span><span class="co">## [3,] -2.3845440 0.08795112  1.16277476  0.2721857  0.28169183  0.4604845</span></span>
<span><span class="co">## [4,] -2.2187255 0.65176353  0.20339372  1.2128839  1.07096226  0.4317445</span></span>
<span><span class="co">## [5,] -0.5669189 1.84121180  0.06402206 -0.3221178  0.02488601  0.3517457</span></span>
<span><span class="co">## [6,] -1.0554684 1.07326688 -0.42296176 -0.3972216 -0.72847203  0.7158252</span></span>
<span><span class="co">##            [,13]      [,14]        [,15]       [,16]      [,17]       [,18]</span></span>
<span><span class="co">## [1,]  0.22645766  0.7414184  0.885021243  1.74928451  0.4172434 -0.36172912</span></span>
<span><span class="co">## [2,] -0.11743827 -1.3097013  0.335409580 -0.19737044  1.9184451  0.06779531</span></span>
<span><span class="co">## [3,] -0.09524678 -0.7460559 -0.583083970  0.25165178  0.6701605  0.43165415</span></span>
<span><span class="co">## [4,] -0.04954717 -1.0375261 -0.005446275 -0.59153455  0.4787165  0.06904808</span></span>
<span><span class="co">## [5,]  1.03117637 -0.4426845 -0.023429569  0.03480782 -0.1772584 -0.21056960</span></span>
<span><span class="co">## [6,]  0.35505985 -0.2426742  1.098536792  0.38752300  0.1341204  0.42348720</span></span>
<span><span class="co">##           [,19]       [,20]</span></span>
<span><span class="co">## [1,]  0.2994094  0.71759073</span></span>
<span><span class="co">## [2,] -0.3500336 -0.06717946</span></span>
<span><span class="co">## [3,]  0.6338821  0.32693292</span></span>
<span><span class="co">## [4,]  0.6517039  0.23217160</span></span>
<span><span class="co">## [5,] -1.2366838  0.41684787</span></span>
<span><span class="co">## [6,] -0.9978965  0.58964001</span></span></code></pre>
<p>This low-dimensional gene expression data can then be used for any
later tasks which would benefit from a low-dimensional embedding of the
gene expression data for all the samples, rather than just the gene
expression data.</p>
</div>
<div class="section level3">
<h3 id="multi-sample-spatial-dataset-single-sample-single-cell-dataset">Multi-sample spatial dataset, single-sample single-cell dataset<a class="anchor" aria-label="anchor" href="#multi-sample-spatial-dataset-single-sample-single-cell-dataset"></a>
</h3>
<p>Here, we perform wSIR on samples 1 and 2 together. This requires the
gene expression matrices from both samples joined together into one
matrix, and the coords dataframes joined into one dataframe. We do this
concatenation using rbind. We then specify the sample IDs for each row
in the joined expression matrix and coordinates dataframe using the
samples argument. This is a vector with a “1” for each row in sample 1
and a “2” for each row in sample 2.</p>
<p>We use the resulting wSIR output to project the gene expression data
from sample 3 into low-dimensional space. We check the dimension of the
resulting matrix.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wsir_obj_samples12</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/wSIR.html">wSIR</a></span><span class="op">(</span>X <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">sample1_exprs</span>, <span class="va">sample2_exprs</span><span class="op">)</span>,</span>
<span>                           coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">sample1_coords</span>, <span class="va">sample2_coords</span><span class="op">)</span>,</span>
<span>                           samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sample1_coords</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sample2_coords</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                           slices <span class="op">=</span> <span class="va">optim_obj</span><span class="op">$</span><span class="va">best_slices</span>, </span>
<span>                           alpha <span class="op">=</span> <span class="va">optim_obj</span><span class="op">$</span><span class="va">best_alpha</span>,</span>
<span>                           optim_params <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sample3_low_dim_exprs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/projectWSIR.html">projectWSIR</a></span><span class="op">(</span>wsir <span class="op">=</span> <span class="va">wsir_obj_samples12</span>, newdata <span class="op">=</span> <span class="va">sample3_exprs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">sample3_low_dim_exprs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 4607   29</span></span></code></pre>
<p>This low-dimensional matrix can then be used for downstream tasks
which would benefit from a low-dimensional embedding of sample 3’s gene
expression matrix that contains information about each cell’s location.
Examples for downstream use include spatial alignment of single-cell and
spatial gene expression data via <a href="https://www.nature.com/articles/s41592-021-01264-7" class="external-link">Tangram</a>,
using the wSIR scores as the input rather than the (unreduced) gene
expression matrix.</p>
<p>An example of a very simple application is using the wSIR scores as
an input to a KNN cell type classification algorithm. This is
demonstrated below, using the ‘knn’ function from ‘class’ package.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samples12_cell_types</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/append.html" class="external-link">append</a></span><span class="op">(</span><span class="va">sample1_cell_types</span>, <span class="va">sample2_cell_types</span><span class="op">)</span></span>
<span></span>
<span><span class="va">knn_classification_object</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Rfast/man/knn.html" class="external-link">knn</a></span><span class="op">(</span>train <span class="op">=</span> <span class="va">wsir_obj_samples12</span><span class="op">$</span><span class="va">scores</span>, </span>
<span>                                test <span class="op">=</span> <span class="va">sample3_low_dim_exprs</span>,</span>
<span>                                cl <span class="op">=</span> <span class="va">samples12_cell_types</span>,</span>
<span>                                k <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">knn_classification_object</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] Presomitic mesoderm Presomitic mesoderm Presomitic mesoderm</span></span>
<span><span class="co">## [4] Presomitic mesoderm Presomitic mesoderm Presomitic mesoderm</span></span>
<span><span class="co">## 24 Levels: Allantois Anterior somitic tissues ... Surface ectoderm</span></span></code></pre>
<p>In the above code, we use the wSIR scores as input for a simple
KNN-based cell type classification tool. We print the tail of the
prediction vector, demonstrating how we can use the wSIR-based
low-dimensional gene expression data from a new sample as a step in a
realistic analysis pipeline.</p>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] class_7.3-22       umap_0.2.10.0      vctrs_0.6.5        doBy_4.6.22       </span></span>
<span><span class="co">##  [5] Rfast_2.1.0        RcppParallel_5.1.9 RcppZiggurat_0.1.6 Rcpp_1.0.13       </span></span>
<span><span class="co">##  [9] ggplot2_3.5.1      magrittr_2.0.3     wSIR_0.99.0        BiocStyle_2.32.1  </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] gtable_0.3.5         xfun_0.47            bslib_0.8.0         </span></span>
<span><span class="co">##  [4] lattice_0.22-6       tools_4.4.1          generics_0.1.3      </span></span>
<span><span class="co">##  [7] parallel_4.4.1       tibble_3.2.1         fansi_1.0.6         </span></span>
<span><span class="co">## [10] highr_0.11           pkgconfig_2.0.3      Matrix_1.7-0        </span></span>
<span><span class="co">## [13] desc_1.4.3           lifecycle_1.0.4      farver_2.1.2        </span></span>
<span><span class="co">## [16] compiler_4.4.1       stringr_1.5.1        textshaping_0.4.0   </span></span>
<span><span class="co">## [19] microbenchmark_1.5.0 munsell_0.5.1        codetools_0.2-20    </span></span>
<span><span class="co">## [22] htmltools_0.5.8.1    sass_0.4.9           yaml_2.3.10         </span></span>
<span><span class="co">## [25] pillar_1.9.0         pkgdown_2.1.1        jquerylib_0.1.4     </span></span>
<span><span class="co">## [28] tidyr_1.3.1          MASS_7.3-60.2        openssl_2.2.1       </span></span>
<span><span class="co">## [31] BiocParallel_1.38.0  cachem_1.1.0         boot_1.3-30         </span></span>
<span><span class="co">## [34] RSpectra_0.16-2      Deriv_4.1.6          tidyselect_1.2.1    </span></span>
<span><span class="co">## [37] digest_0.6.37        stringi_1.8.4        dplyr_1.1.4         </span></span>
<span><span class="co">## [40] purrr_1.0.2          bookdown_0.40        labeling_0.4.3      </span></span>
<span><span class="co">## [43] distances_0.1.11     cowplot_1.1.3        fastmap_1.2.0       </span></span>
<span><span class="co">## [46] grid_4.4.1           colorspace_2.1-1     cli_3.6.3           </span></span>
<span><span class="co">## [49] utf8_1.2.4           broom_1.0.6          withr_3.0.1         </span></span>
<span><span class="co">## [52] scales_1.3.0         backports_1.5.0      rmarkdown_2.28      </span></span>
<span><span class="co">## [55] modelr_0.1.11        reticulate_1.39.0    png_0.1-8           </span></span>
<span><span class="co">## [58] askpass_1.2.0        ragg_1.3.3           evaluate_1.0.0      </span></span>
<span><span class="co">## [61] knitr_1.48           rlang_1.1.4          glue_1.7.0          </span></span>
<span><span class="co">## [64] BiocManager_1.30.25  jsonlite_1.8.8       R6_2.5.1            </span></span>
<span><span class="co">## [67] systemfonts_1.1.0    fs_1.6.4</span></span></code></pre>
</details>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Max Woollard, Linh Nghiem, Shila Ghazanfar.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
